src/context/CategoryContext.jsx

```diff
@@ -0,0 +1,201 @@
+      const result = await callApi("post", "categories", false, category);
+      dispatchCategories({
+        type: ACTION_TYPE_SUCCESS,
+        payload: [...categories.payload, result.data.category],
+      });
+    } catch (err) {
+      dispatchCategories({
+        type: ACTION_TYPE_FAILURE,
+        payload: formatError(err),
+      });
+    }
+  };
+
+  const deleteCategory = async (categoryId) => {
+    dispatchCategories({ type: ACTION_TYPE_LOADING });
+    try {
+      await callApi("delete", `categories/${categoryId}`, false);
+      dispatchCategories({
+        type: ACTION_TYPE_SUCCESS,
+        payload: categories.payload.filter(
+          (category) => category.id !== categoryId
+        ),
+      });
+    } catch (err) {
+      dispatchCategories({
+        type: ACTION_TYPE_FAILURE,
+        payload: formatError(err),
+      });
+    }
+  };
+  const updateCategory = async (categoryId, category) => {
+    dispatchCategories({ type: ACTION_TYPE_LOADING });
+    try {
+      const result = await callApi(
+        "put",
+        `categories/${categoryId}`,
+        false,
+        category
+      );
+      dispatchCategories({
+        type: ACTION_TYPE_SUCCESS,
+        payload: categories.payload.map((category) =>
+          category.id === categoryId ? result.data.category : category
+        ),
+      });
+    } catch (err) {
+      dispatchCategories({
+        type: ACTION_TYPE_FAILURE,
+        payload: formatError(err),
+      });
+    }
+  };
+  const fetchYourCategories = async () => {
+    dispatchYourCategories({ type: ACTION_TYPE_LOADING });
+    try {
+      const result = await callApi("get", "user/categories", true);
+      dispatchYourCategories({
+        type: ACTION_TYPE_SUCCESS,
+        payload: result.data.categories,
+      });
+    } catch (err) {
+      dispatchYourCategories({
+        type: ACTION_TYPE_FAILURE,
+        payload: formatError(err),
+      });
+    }
+  };
+  const searchYourCategories = async (search) => {
+    dispatchYourCategories({ type: ACTION_TYPE_LOADING });
+    try {
+      const result = await callApi(
+        "get",
+        `user/categories?search=${search}`,
+        true
+      );
+      dispatchYourCategories({
+        type: ACTION_TYPE_SUCCESS,
+        payload: result.data.categories,
+      });
+    } catch (err) {
+      dispatchYourCategories({
+        type: ACTION_TYPE_FAILURE,
+        payload: formatError(err),
+      });
+    }
+  };
+  const searchCategories = async (search) => {
+    dispatchCategories({ type: ACTION_TYPE_LOADING });
+    try {
+      dispatchCategories({ type: ACTION_TYPE_SUCCESS, payload: [] });
+      const result = await callApi("get", `categories?search=${search}`, false);
+      dispatchCategories({
+        type: ACTION_TYPE_SUCCESS,
+        payload: result.data.categories,
+      });
+    } catch (err) {
+      dispatchCategories({
+        type: ACTION_TYPE_FAILURE,
+        payload: formatError(err),
+      });
+    }
+  };
```

Comment:
```dispatchYourCategories({ type: ACTION_TYPE_LOADING });
    try {
      const result = await callApi(
        "get",
        `user/categories?search=${search}`,
        true
      );
      dispatchYourCategories({
        type: ACTION_TYPE_SUCCESS,
        payload: result.data.categories,
      });
    } catch (err) {
      dispatchYourCategories({
        type: ACTION_TYPE_FAILURE,
        payload: formatError(err),
      });
    }
    ```
    
 This code part is same repeated in many functions  except for the **ActionType**, **payload** and **callApi** arguments,  maybe use a generic function that takes these parameters and call required function.